aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
```

Comment:
Would recommend:
```
Fields with an effective value must be expressed as separate fields in the API:
* mutable field(s) configured by the user
* output only field(s) that reveal the effective value(s) which resulted
```

for specific cases

```
string ip_address;
string effective_ip_address [(google.api.field_behavior) = OUTPUT_ONLY];
```
In this case you would say the ip_address empty-value "" is defined meaning dynamically allocated. You would also say the default value of ip_address is the "" empty-value.

An API designer even plan in advance and say that the ip_address string "dynamic" means dynamically allocated, and that the default is "dynamic". 

```
string ip_address [(google.api.default) = "dynamic"];
string effective_ip_address [(google.api.field_behavior) = OUTPUT_ONLY];
->
POST { ... provide no opinion ...}
GET { "ip_address": "dynamic", "effective_ip_address": "10.5.29.234" }
```

That would give you even more precise resource response output, and enable different kinds of non-literal IP allocation, e.g. ip_address = "dynamic:10.5.0.0/16"

a real-world example of that would be in the hosted-software-version field where the default value of the setting means "server's choice", but there's also "server's choice of cutting edge version" (release channel).

```
string minecraft_version [(google.api.default) = "stable"];
string effective_minecraft_version [(google.api.field_behavior) = OUTPUT_ONLY];
---
POST { ... provide no opinion ... }
GET { "minecraft_version": "stable", "effective_minecraft_version": "3.45" }
---
POST { "minecraft_version": "2.67" }
GET { "minecraft_version": "2.67", "effective_minecraft_version": "2.67" }
---
POST { "minecraft_version": "unstable" }
GET { "minecraft_version": "unstable", "effective_minecraft_version": "3.89-preview.2" }
```

Reply:
Modified. Do we want to show an example of how to use an automatic updating version tag? I think it's unclear how a "version": "latest" will resolve as the latest version changes over time. Does the server automatically upgrade to keep you on the latest, or is it a one-time thing? We need some clear guidance here to avoid contrasting implementations